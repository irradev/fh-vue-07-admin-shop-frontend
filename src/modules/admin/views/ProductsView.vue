<template>
  <div class="bg-white px-5 py-2 rounded">
    <h1 class="text-3xl text-slate-700">Productos</h1>
    <!-- component -->
    <div class="py-8 w-full">
      <div class="shadow overflow-hidden rounded border-b border-gray-200">
        <table class="min-w-full bg-white">
          <thead class="bg-gray-800 text-white">
            <tr>
              <th class="w-20 text-left py-3 px-4 uppercase font-semibold text-sm">Imagen</th>
              <th class="flex-1 text-left py-3 px-4 uppercase font-semibold text-sm">TÃ­tulo</th>
              <th class="text-left py-3 px-4 uppercase font-semibold text-sm">Precio</th>
              <th class="text-left py-3 px-4 uppercase font-semibold text-sm">Tallas</th>
            </tr>
          </thead>
          <tbody class="text-gray-700">
            <tr
              v-for="(product, index) in products"
              :key="product.id"
              :class="{
                'bg-gray-200': index % 2 === 0,
              }"
            >
              <td class="text-left py-3 px-4">
                <img
                  :src="product.images[0]"
                  :alt="`photo of ${product.title}`"
                  class="w-10 h-10 object-cover"
                />
              </td>
              <td class="text-left py-3 px-4">
                <RouterLink
                  :to="`products/${product.id}`"
                  class="hover:text-blue-500 hover:underline"
                >
                  {{ product.title }}
                </RouterLink>
              </td>
              <td class="text-left py-3 px-4">
                <span
                  class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"
                >
                  $ {{ product.price }}
                </span>
              </td>
              <td class="text-left py-3 px-4">
                {{ product.sizes.join(', ') }}
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>

  <BottomPagination :is-no-more-data="!!products && products.length < 10" :page="page" />
</template>

<script setup lang="ts">
  import { watchEffect } from 'vue';
  import { useQuery, useQueryClient } from '@tanstack/vue-query';
  import { getProductsAction } from '@/modules/products/actions';
  import BottomPagination from '@/modules/common/components/ButtonPagination.vue';
  import { usePagination } from '@/modules/common/composables/usePagination';

  const queryClient = useQueryClient();
  const { page } = usePagination();

  const { data: products, isLoading } = useQuery({
    queryKey: ['products', { page: page }],
    queryFn: () => getProductsAction(page.value),
    staleTime: 60 * 1000,
  });

  watchEffect(() => {
    queryClient.prefetchQuery({
      queryKey: ['products', { page: page.value + 1 }],
      queryFn: () => getProductsAction(page.value + 1),
    });
  });
</script>
