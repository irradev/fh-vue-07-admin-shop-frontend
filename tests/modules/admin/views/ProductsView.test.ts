import { shallowMount } from '@vue/test-utils';
import MockAdapter from 'axios-mock-adapter';
import { tesloApi } from '@/api/tesloApi';
import ProductsView from '@/modules/admin/views/ProductsView.vue';
import { fakeProducts } from '../../../_mockData/fakeProducts';
import type { Mock } from 'vitest';
import { useQuery, useQueryClient } from '@tanstack/vue-query';
import { createRouter, createWebHistory } from 'vue-router';

const tesloApiMock = new MockAdapter(tesloApi);

tesloApiMock.onGet('/products?limit=10&offset=0').reply(200, []);
tesloApiMock.onGet('/products?limit=10&offset=10').reply(200, []);

vi.mock('@tanstack/vue-query', () => {
  return {
    useQueryClient: vi.fn().mockReturnValue({ prefetchQuery: vi.fn() }),
    useQuery: vi.fn(),
  };
});

const router = createRouter({
  history: createWebHistory(),
  routes: [
    {
      path: '',
      component: ProductsView,
    },
  ],
});

describe('<ProductsView />', () => {
  (useQuery as Mock).mockReturnValue({
    data: fakeProducts,
    isLoading: false,
    isError: false,
  });

  window.scrollTo = vi.fn();

  const wrapper = shallowMount(ProductsView, {
    global: {
      plugins: [router],
    },
  });

  test('should render with default values', async () => {
    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should prefetch query on mounted', async () => {
    await router.replace('/?page=2');

    expect(window.scrollTo).toHaveBeenCalledWith({
      behavior: 'smooth',
      top: 0,
    });

    expect(useQueryClient().prefetchQuery).toHaveBeenCalledWith({
      queryKey: ['products', { page: 2 }],
      queryFn: expect.any(Function),
    });
  });
});
