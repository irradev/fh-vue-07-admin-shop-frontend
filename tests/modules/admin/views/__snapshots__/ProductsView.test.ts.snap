// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ProductsView /> > should render with default values 1`] = `
"<div class="bg-white px-5 py-2 rounded">
  <h1 class="text-3xl text-slate-700">Productos</h1><!-- component -->
  <div class="py-8 w-full">
    <div class="shadow overflow-hidden rounded border-b border-gray-200">
      <table class="min-w-full bg-white">
        <thead class="bg-gray-800 text-white">
          <tr>
            <th class="w-20 text-left py-3 px-4 uppercase font-semibold text-sm">Imagen</th>
            <th class="flex-1 text-left py-3 px-4 uppercase font-semibold text-sm">Título</th>
            <th class="text-left py-3 px-4 uppercase font-semibold text-sm">Precio</th>
            <th class="text-left py-3 px-4 uppercase font-semibold text-sm">Tallas</th>
          </tr>
        </thead>
        <tbody class="text-gray-700">
          <tr class="bg-gray-200">
            <td class="text-left py-3 px-4"><img src="1740176-00-A_0_2000.jpg" alt="photo of Men’s Chill Crew Neck Sweatshirt" class="w-10 h-10 object-cover"></td>
            <td class="text-left py-3 px-4">
              <router-link-stub to="products/008ea87d-a16c-4c25-9de3-8a14bae089cd" replace="false" custom="false" ariacurrentvalue="page" class="hover:text-blue-500 hover:underline"></router-link-stub>
            </td>
            <td class="text-left py-3 px-4"><span class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"> $ 75</span></td>
            <td class="text-left py-3 px-4">XS, S, L, XL, XXL</td>
          </tr>
          <tr class="">
            <td class="text-left py-3 px-4"><img src="8529107-00-A_0_2000.jpg" alt="photo of 3D Large Wordmark Pullover Hoodie" class="w-10 h-10 object-cover"></td>
            <td class="text-left py-3 px-4">
              <router-link-stub to="products/00caa867-abf7-4172-879b-2c17c8cad1d6" replace="false" custom="false" ariacurrentvalue="page" class="hover:text-blue-500 hover:underline"></router-link-stub>
            </td>
            <td class="text-left py-3 px-4"><span class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"> $ 70</span></td>
            <td class="text-left py-3 px-4">XS, S, XL, XXL</td>
          </tr>
          <tr class="bg-gray-200">
            <td class="text-left py-3 px-4"><img src="1506211-00-A_1_2000.jpg" alt="photo of Kids Corp Jacket" class="w-10 h-10 object-cover"></td>
            <td class="text-left py-3 px-4">
              <router-link-stub to="products/02f68a98-208c-4e2c-8ed3-742e07ba7ef7" replace="false" custom="false" ariacurrentvalue="page" class="hover:text-blue-500 hover:underline"></router-link-stub>
            </td>
            <td class="text-left py-3 px-4"><span class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"> $ 30</span></td>
            <td class="text-left py-3 px-4">XS, S, M</td>
          </tr>
          <tr class="">
            <td class="text-left py-3 px-4"><img src="8764734-00-A_1.jpg" alt="photo of Men's 3D Large Wordmark Tee" class="w-10 h-10 object-cover"></td>
            <td class="text-left py-3 px-4">
              <router-link-stub to="products/081ed246-4a36-43c5-870b-8f687427bf93" replace="false" custom="false" ariacurrentvalue="page" class="hover:text-blue-500 hover:underline"></router-link-stub>
            </td>
            <td class="text-left py-3 px-4"><span class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"> $ 35</span></td>
            <td class="text-left py-3 px-4">XS, S, M</td>
          </tr>
          <tr class="bg-gray-200">
            <td class="text-left py-3 px-4"><img src="8765090-00-A_0_2000.jpg" alt="photo of Women's T Logo Short Sleeve Scoop Neck Tee" class="w-10 h-10 object-cover"></td>
            <td class="text-left py-3 px-4">
              <router-link-stub to="products/0beeea8d-23c5-4ea0-a1d0-ebb7eb7a2aa5" replace="false" custom="false" ariacurrentvalue="page" class="hover:text-blue-500 hover:underline"></router-link-stub>
            </td>
            <td class="text-left py-3 px-4"><span class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"> $ 35</span></td>
            <td class="text-left py-3 px-4">XS, S, M, L, XL, XXL</td>
          </tr>
          <tr class="">
            <td class="text-left py-3 px-4"><img src="1740250-00-A_1.jpg" alt="photo of Men's Raven Lightweight Zip Up Bomber Jacket" class="w-10 h-10 object-cover"></td>
            <td class="text-left py-3 px-4">
              <router-link-stub to="products/0c8e0838-e0b6-45e2-9479-b2e882d2331b" replace="false" custom="false" ariacurrentvalue="page" class="hover:text-blue-500 hover:underline"></router-link-stub>
            </td>
            <td class="text-left py-3 px-4"><span class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"> $ 130</span></td>
            <td class="text-left py-3 px-4">S, M, L, XL, XXL</td>
          </tr>
          <tr class="bg-gray-200">
            <td class="text-left py-3 px-4"><img src="8765120-00-A_0_2000.jpg" alt="photo of Women's Small Wordmark Short Sleeve V-Neck Tee" class="w-10 h-10 object-cover"></td>
            <td class="text-left py-3 px-4">
              <router-link-stub to="products/0df45604-0309-4410-9bc3-9e681d2566a9" replace="false" custom="false" ariacurrentvalue="page" class="hover:text-blue-500 hover:underline"></router-link-stub>
            </td>
            <td class="text-left py-3 px-4"><span class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"> $ 35</span></td>
            <td class="text-left py-3 px-4">XS, S, M, L, XL, XXL</td>
          </tr>
          <tr class="">
            <td class="text-left py-3 px-4"><img src="8529342-00-A_1.jpg" alt="photo of Kids Cybertruck Tee" class="w-10 h-10 object-cover"></td>
            <td class="text-left py-3 px-4">
              <router-link-stub to="products/1283ce77-2be8-49c1-95ac-0f307108a8bb" replace="false" custom="false" ariacurrentvalue="page" class="hover:text-blue-500 hover:underline"></router-link-stub>
            </td>
            <td class="text-left py-3 px-4"><span class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"> $ 25</span></td>
            <td class="text-left py-3 px-4">XS, S, M</td>
          </tr>
          <tr class="bg-gray-200">
            <td class="text-left py-3 px-4"><img src="1703767-00-A_1.jpg" alt="photo of Men's Solar Roof Tee" class="w-10 h-10 object-cover"></td>
            <td class="text-left py-3 px-4">
              <router-link-stub to="products/12ad7d60-45b8-49a8-92a4-f4db225f39f6" replace="false" custom="false" ariacurrentvalue="page" class="hover:text-blue-500 hover:underline"></router-link-stub>
            </td>
            <td class="text-left py-3 px-4"><span class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"> $ 35</span></td>
            <td class="text-left py-3 px-4">S, M, L, XL</td>
          </tr>
          <tr class="">
            <td class="text-left py-3 px-4"><img src="1549275-00-A_1.jpg" alt="photo of Women's Plaid Mode Tee" class="w-10 h-10 object-cover"></td>
            <td class="text-left py-3 px-4">
              <router-link-stub to="products/1560071e-96fe-41bb-ae70-6c6c72774486" replace="false" custom="false" ariacurrentvalue="page" class="hover:text-blue-500 hover:underline"></router-link-stub>
            </td>
            <td class="text-left py-3 px-4"><span class="bg-blue-200 text-blue-600 py-1 px-3 rounded-full text-xs whitespace-nowrap"> $ 35</span></td>
            <td class="text-left py-3 px-4">S, M</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
<bottom-pagination-stub page="1" isnomoredata="false"></bottom-pagination-stub>"
`;
